libbeat/processors/add_host_metadata/add_host_metadata_test.go

```diff
@@ -522,6 +523,82 @@ func TestFQDNEventSync(t *testing.T) {
 	}, time.Second*3600, time.Millisecond*10)
 }
 
+func TestDataReload(t *testing.T) {
+	testConfig := conf.MustNewConfigFrom(map[string]interface{}{
+		"cache.ttl": "5m",
+	})
+
+	// Start with FQDN off
+	err := features.UpdateFromConfig(conf.MustNewConfigFrom(map[string]interface{}{
+		"features.fqdn.enabled": false,
+	}))
+	require.NoError(t, err)
+
+	info := &mockHostInfo{}
+	factory := func() (hostInfo, error) {
+		return info, nil
+	}
+
+	p, err := newWithHostInfoFactory(testConfig, logptest.NewTestingLogger(t, ""), factory)
+	require.NoError(t, err)
+
+	// we should have a single data reload during creation
+	assert.Equal(t, 1, info.HostInfoRequestCount)
+	assert.Equal(t, 0, info.FQDNRequestCount)
+
+	closeCh := make(chan struct{})
+	wg := &sync.WaitGroup{}
+	t.Cleanup(func() {
+		close(closeCh)
+		wg.Wait()
+	})
+	// start some goroutines enriching events in an infinite loop
+	for range 10 {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				select {
+				case <-closeCh:
+					return
+				default:
+				}
+				_, err := p.Run(&beat.Event{
+					Fields: mapstr.M{},
+				})
+				require.NoError(t, err)
+			}
+		}()
+	}
+
+	// update
+	err = features.UpdateFromConfig(conf.MustNewConfigFrom(map[string]interface{}{
```

Comment:
We don't control or know how many executions of the goroutines have happened when the configuration changes happen. This makes this test less deterministic than I would like. There is a slim chance none of the goroutines have started yet when you make this configuration change for example.

It is slightly more work but you can synchronize on ensuring the goroutines have done some executions before+after each config change.

Reply:
We can't deterministically test the usage of atomics (we can't guarantee two goroutines cause a conflict in the CAS operation easily or at all), I think in the original bug there was no dependency on concurrency it was just a logic bug.

So you could simplify this to not use goroutines and just call `p.Run` a few times deterministically in the test to make sure the transitions happen correctly.

Reply:
I see that point I am just a bit uneasy with an `assert.Equal(collect, 1, info.FQDNRequestCount)` that isn't paired with a check that the goroutines executed more times than that. They very, very likely did but we don't guarantee it.

I can buy into not significantly rewriting the test to execute serially if you don't think it helps anything.

Reply:
Added a simple event counter here: https://github.com/elastic/beats/pull/46546/commits/b501156548c34a271fbeee3720971a6a1a31238c. We now wait until we process a number of events equal to the number of goroutines, which is a bit arbitrary, but should work well enough.